File: Data Structures & Algorithms/valid-tree/submission-4.py
class Solution:
    def validTree(self, n: int, edges: List[List[int]]) -> bool:
        visit = set()

        adj = {i: [] for i in range(n)}

        for i,j in edges:
            adj[i].append(j)
            adj[j].append(i)
        
        prev=-1
        
        def dfs(node,prev):
            if node in visit:
                return False

            visit.add(node)

            
            
            
            for nei in adj[node]:
                if nei == prev:
                    continue
                if not dfs(nei, node):
                    return False
            return True
            
        
            
            
        

        return dfs(0,-1) and len(visit) == n

                

            
            

File: Data Structures & Algorithms/climbing-stairs/submission-4.java
class Solution {
    public int climbStairs(int n) {
        int[] memo = new int[n+1];
        for(int i = 0; i< n+1; i++){
            memo[i] = -1;
        }
        return dfs(0,n, memo);
        
    }

    public int dfs(int i, int n, int[] memo){
        if(i == n){
            return 1;
        }
        if(i>n){
            return 0;
        }
        if(memo[i] != -1){
            return memo[i];
        }

        int res = dfs(i+1,n, memo) + dfs(i+2, n, memo);
        memo[i] = res;
        return res;

        
    }
}
